common/src/test/java/org/conscrypt/MacTest.java

```diff
@@ -0,0 +1,191 @@
+                    expectedBytes, macBytes);
+
+            // Calculated using a pre-loved Mac
+            macBytes = generateReusingMac(algorithm, keyBytes, msgBytes);
+            assertArrayEquals(failMessage("Re-use Mac", baseFailMsg, macBytes),
+                    expectedBytes, macBytes);
+        }
+    }
+
+    private String failMessage(String test, String base, byte[] mac) {
+        return String.format("Test %s\n%s\nActual=  %s", test, base, encodeHex(mac));
+    }
+
+    // Splits a ByteBuffer into an array of NUM_SPLITS ByteBuffers containing the same data.
+    // If input.remaining < NUM_SPLITS then some buffers will be empty, which is fine.
+    private ByteBuffer[] split(ByteBuffer input) {
+        ByteBuffer[] buffers = new ByteBuffer[NUM_SPLITS];
+        int targetSize = (input.remaining() / NUM_SPLITS) + 1;
+        ByteBuffer buffer;
+        for (int i = 0; i < NUM_SPLITS; i++) {
+            int size = Math.min(targetSize, input.remaining());
+            buffer = input.isDirect() ? ByteBuffer.allocateDirect(size) : ByteBuffer.allocate(size);
+            buffers[i] = buffer;
+
+            int savedLimit = input.limit();
+            input.limit(input.position() + size);
+            buffer.put(input);
+            buffer.flip();
+            input.limit(savedLimit);
+        }
+        assertEquals(0, input.remaining());
+        return buffers;
+    }
+
+    private byte[] generateMacUsingUpdate(String algorithm, SecretKeySpec key, byte[] message)
+            throws Exception {
+        Mac mac = Mac.getInstance(algorithm);
+        assertNotNull(mac);
+        mac.init(key);
+        mac.update(message);
+        return mac.doFinal();
+    }
+
+    private byte[] generateMacUsingFinal(String algorithm, SecretKeySpec key, byte[] message)
+            throws Exception {
+        Mac mac = Mac.getInstance(algorithm);
+        assertNotNull(mac);
+        mac.init(key);
+        return mac.doFinal(message);
+    }
+
+    private byte[] generateMac(String algorithm, SecretKeySpec key, ByteBuffer buffer)
+            throws Exception {
+        return generateMac(algorithm, key, new ByteBuffer[] { buffer });
+    }
+
+    private byte[] generateMac(String algorithm, SecretKeySpec key, ByteBuffer[] buffers)
+            throws Exception {
+        Mac mac = Mac.getInstance(algorithm);
+        assertNotNull(mac);
+        mac.init(key);
+        for (ByteBuffer buffer : buffers) {
+            mac.update(buffer);
+        }
+        return mac.doFinal();
+    }
+
+    private byte[] generateReusingMac(String algorithm, byte[] keyBytes, byte[] message)
+            throws Exception {
+        Mac mac = Mac.getInstance(algorithm);
+        assertNotNull(mac);
+
+        // Mutate the original message and key and calculate a MAC from them
+        byte[] otherKeyBytes = Arrays.copyOf(keyBytes, keyBytes.length);
+        Arrays.sort(otherKeyBytes);
```

Comment:
why do we sort here? is this to make the test deterministic?

Reply:
Just a cheap way of mutating the key and message whilst keeping the same lengths.  Maybe it would be clearer if I used `SecureRandom.nextBytes()`.

Reply:
Done... Also turns out I forgot to commit a tidyup CL from yesterday.